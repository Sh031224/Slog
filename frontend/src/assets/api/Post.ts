import axios from "axios";
import { server } from "../../config/index.json";

interface PostParmsType {
  page: number;
  limit: number;
  order?: string;
  category?: number;
}

interface CreateTempPostType {
  title: string;
  description: string | null;
  content: string | null;
  thumbnail: string | null;
  category_idx: number | null;
}

interface CreatePostType {
  title: string;
  description: string;
  content: string;
  thumbnail: string | null;
  category_idx: number;
}

interface ModifyPostType {
  title: string;
  description: string | null;
  content: string | null;
  thumbnail: string | null;
  category_idx?: number;
  is_temp?: boolean;
}

class PostList {
  async GetTempPosts() {
    let url = `${server}/api/v1/post/temp`;

    const { data } = await axios.get(url);

    return data;
  }

  async GetPostList(query: PostParmsType) {
    let url = `${server}/api/v1/post/?page=${query.page}&limit=${query.limit}`;

    if (query.order) url += `&order=${query.order}`;
    if (query.category) url += `&category=${query.category}`;
    const { data } = await axios.get(url);

    return data;
  }

  async GetPostInfo(idx: number) {
    const url = `${server}/api/v1/post/${idx}`;

    const { data } = await axios.get(url);

    return data;
  }

  async GetPostCommentCount(idx: number) {
    const url = `${server}/api/v1/post/comment/${idx}`;

    const { data } = await axios.get(url);

    return data;
  }

  async GetPostSearch(query: string) {
    const url = `${server}/api/v1/post/search?query=${query}`;

    const { data } = await axios.get(url);

    return data;
  }

  async UploadFiles(files: File[]) {
    const url = `${server}/api/v1/upload`;

    const formData = new FormData();
    formData.append("files", files[0]);

    const { data } = await axios.post(url, formData);

    return data;
  }

  async CreateTempPost(body: CreateTempPostType) {
    const url = `${server}/api/v1/post/temp`;

    const { data } = await axios.post(url, body);

    return data;
  }

  async CreatePost(body: CreatePostType) {
    const url = `${server}/api/v1/post`;

    const { data } = await axios.post(url, body);

    return data;
  }

  async ModifyPost(post_idx: number, body: ModifyPostType) {
    const url = `${server}/api/v1/post/${post_idx}`;

    const { data } = await axios.put(url, body);

    return data;
  }

  async DeletePost(post_idx: number) {
    const url = `${server}/api/v1/post/${post_idx}`;

    const { data } = await axios.delete(url);

    return data;
  }
}

export default new PostList();
