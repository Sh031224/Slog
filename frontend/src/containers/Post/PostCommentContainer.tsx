import PostCommentItem from "components/Post/PostComment/PostCommentItem";
import React, { useCallback, useState } from "react";
import CommentType from "types/CommentType";
import { GetRepliesResponse } from "types/Response";

interface PostCommentContainerProps {
  comment: CommentType;
  admin: boolean;
  getReplies: (comment_idx: number) => Promise<GetRepliesResponse>;
  userId: number;
  modifyComment: (comment_idx: number, content: string) => Promise<void>;
  deleteComment: (comment_idx: number) => void;
  login: boolean;
  createReply: (
    comment_idx: number,
    content: string,
    is_private?: boolean | undefined
  ) => Promise<void>;
  modifyReply: (reply_idx: number, content: string) => Promise<void>;
  deleteReply: (reply_idx: number) => void;
}

const PostCommentContainer = ({
  comment,
  admin,
  getReplies,
  userId,
  modifyComment,
  deleteComment,
  login,
  createReply,
  modifyReply,
  deleteReply
}: PostCommentContainerProps) => {
  const [modify, setModify] = useState<boolean>(false);
  const [modifyInput, setModifyInput] = useState<string>(comment.content);
  const [reply, setReply] = useState<boolean>(false);
  const [replyInput, setReplyInput] = useState<string>("");
  const [isPrivate, setIsPrivate] = useState<boolean>(comment.is_private);

  const cancelModify = useCallback(() => {
    setModify(false);
    setModifyInput(comment.content);
  }, [comment]);

  const cancelReply = useCallback(() => {
    setReply(false);
    setReplyInput("");
  }, []);

  const setIsPrivateCallback = useCallback(
    (status: boolean) => {
      if (comment.is_private) {
        setIsPrivate(true);
      } else {
        setIsPrivate(status);
      }
    },
    [comment]
  );

  return (
    <>
      <PostCommentItem
        reply={reply}
        cancelReply={cancelReply}
        setReply={setReply}
        replyInput={replyInput}
        setReplyInput={setReplyInput}
        isPrivate={isPrivate}
        setIsPrivateCallback={setIsPrivateCallback}
        createReply={createReply}
        modifyReply={modifyReply}
        deleteReply={deleteReply}
        login={login}
        modify={modify}
        setModify={setModify}
        modifyInput={modifyInput}
        setModifyInput={setModifyInput}
        cancelModify={cancelModify}
        deleteComment={deleteComment}
        modifyComment={modifyComment}
        userId={userId}
        getReplies={getReplies}
        comment={comment}
        admin={admin}
      />
    </>
  );
};

export default PostCommentContainer;
