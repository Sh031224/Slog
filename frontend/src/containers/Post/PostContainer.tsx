import { inject, observer } from "mobx-react";
import React, { useCallback, useEffect, useState, SetStateAction } from "react";
import { useRouter } from "next/router";
import cookies from "js-cookie";
import { NotificationManager } from "react-notifications";
import { confirmAlert } from "react-confirm-alert";
import Head from "next/head";
import dynamic from "next/dynamic";
import useStore from "lib/hooks/useStore";
import { PostInfoType, PostParmsType } from "types/PostType";
import { GetPostCommentCountResponse, GetPostInfoResponse, ResponseType } from "types/Response";

const Post = dynamic(() => import("components/Post"));

interface PostContainerProps {
  post: PostInfoType;
}

const PostContainer = ({ post }: PostContainerProps) => {
  const router = useRouter();
  const { store } = useStore();
  const { idx } = router.query;

  const { handlePrevUrl, prevUrl } = store.HistoryStore;
  const { getPostInfo, hitPosts, handleHitPosts, deletePost, getPostCommentCount } = store.PostStore;
  const {
    comments,
    initComments,
    getComments,
    getReplies,
    commentCreate,
    commentModify,
    commentDelete,
    replyCreate,
    replyModify,
    replyDelete
  } = store.CommentStore;
  const { admin, login, userId, handleLoginChange } = store.UserStore;

  const [loading, setLoading] = useState(true);
  const [handler, setHandler] = useState<boolean>(false);
  const [commentCount, setCommentCount] = useState<number>(0);
  const [isSaving, setIsSaving] = useState<boolean>(false);
  const [postInfo, setPostInfo] = useState<PostInfoType | SetStateAction<PostInfoType | any>>({});

  const getPostInfoCallback = useCallback(
    async (idx: number) => {
      if (!isNaN(idx)) {
        await getPostInfo(idx).then((response: GetPostInfoResponse) => {
          setPostInfo(response.data.post);
          setCommentCount(response.data.post.comment_count);
        });
      }
    },
    [idx]
  );

  const getCommentsCallback = useCallback(
    async (postIdx: number) => {
      if (!isNaN(postIdx)) {
        await getComments(postIdx).catch((err: Error) => {
          if (err.message !== "Error: Request failed with status code 404") {
            NotificationManager.error("오류가 발생하였습니다.", "Error");
          }
        });
      }
    },
    [idx, login]
  );

  const getHitPostsCallback = useCallback(async () => {
    const query: PostParmsType = {
      page: 1,
      limit: 5,
      order: "hit"
    };
    await handleHitPosts(query);
  }, [idx]);

  const getAllContent = useCallback(async () => {
    setLoading(true);
    if (!isNaN(Number(idx))) {
      try {
        if (!post.idx) {
          //csr
          await getPostInfoCallback(Number(idx));
        } else {
          // ssr
          setPostInfo(post);
          setCommentCount(post.comment_count);
        }
        await getHitPostsCallback();
        setLoading(false);
      } catch (err) {
        if (err.message === "Error: Request failed with status code 404") {
          NotificationManager.warning("해당 게시글이 없습니다.", "Error");
          router.push("/");
        } else {
          router.push("/");
          NotificationManager.error("오류가 발생하였습니다.", "Error");
        }
      }
    }
  }, [idx]);

  const createComment = useCallback(
    async (postIdx: number, content: string, isPrivate?: boolean) => {
      if (!isSaving) {
        setIsSaving(true);
        try {
          await commentCreate(postIdx, content, isPrivate);
          await getCommentsCallback(postIdx);
          await getPostCommentCount(postIdx).then((res: GetPostCommentCountResponse) => {
            setCommentCount(res.data.total_count);
          });
          setIsSaving(false);
        } catch (err) {
          if (err.message === "Error: Request failed with status code 401") {
            cookies.remove("access_token", { path: "/" });
            handleLoginChange(false);
            NotificationManager.warning("로그인 후 작성가능합니다.", "Error");
          } else {
            NotificationManager.error("오류가 발생하였습니다.", "Error");
          }
          setIsSaving(false);
        }
      }
    },
    [login, isSaving]
  );

  const modifyComment = useCallback(
    async (commentIdx: number, content: string) => {
      if (!isSaving) {
        setIsSaving(true);
        try {
          await commentModify(commentIdx, content).then((res: ResponseType) => {
            if (res.status === 200) {
              NotificationManager.success("댓글을 수정하였습니다.", "Success");
            }
          });
          await getCommentsCallback(Number(idx));
          setIsSaving(false);
        } catch (err) {
          if (err.message === "Error: Request failed with status code 403") {
            NotificationManager.warning("권한이 없습니다.", "Error");
          } else if (err.message === "Error: Request failed with status code 401") {
            cookies.remove("access_token", { path: "/" });
            handleLoginChange(false);
            NotificationManager.warning("로그인 후 작성가능합니다.", "Error");
          } else {
            NotificationManager.error("오류가 발생하였습니다.", "Error");
          }
          setIsSaving(false);
        }
      }
    },
    [login, idx]
  );

  const deleteCommentCallback = useCallback(
    async (commentIdx: number) => {
      try {
        await commentDelete(commentIdx).then((res: ResponseType) => {
          if (res.status === 200) {
            NotificationManager.success("댓글을 삭제하였습니다.", "Success");
          }
        });
        await getCommentsCallback(Number(idx));
        await getPostCommentCount(Number(idx)).then((res: GetPostCommentCountResponse) => {
          setCommentCount(res.data.total_count);
        });
      } catch (err) {
        if (err.message === "Error: Request failed with status code 403") {
          NotificationManager.warning("권한이 없습니다.", "Error");
        } else if (err.message === "Error: Request failed with status code 401") {
          cookies.remove("access_token", { path: "/" });
          NotificationManager.warning("로그인 시간이 만료되었습니다.", "Error");
        } else {
          NotificationManager.error("오류가 발생하였습니다.", "Error");
        }
      }
    },
    [login, idx]
  );

  const deleteComment = useCallback(
    async (commentIdx) => {
      confirmAlert({
        title: "Warning",
        message: "정말로 삭제하시겠습니까?",
        buttons: [
          {
            label: "Yes",
            onClick: () => deleteCommentCallback(commentIdx)
          },
          {
            label: "No",
            onClick: () => {
              return;
            }
          }
        ]
      });
    },
    [deleteCommentCallback]
  );

  const createReply = useCallback(
    async (commentIdx: number, content: string, isPrivate?: boolean) => {
      if (!isSaving) {
        setIsSaving(true);
        try {
          await replyCreate(commentIdx, content, isPrivate);
          await getCommentsCallback(Number(idx));
          await getPostCommentCount(Number(idx)).then((res: GetPostCommentCountResponse) => {
            setCommentCount(res.data.total_count);
          });
          setIsSaving(false);
        } catch (err) {
          if (err.message === "Error: Request failed with status code 403") {
            NotificationManager.warning("권한이 없습니다.", "Error");
          } else if (err.message === "Error: Request failed with status code 401") {
            cookies.remove("access_token", { path: "/" });
            handleLoginChange(false);
            NotificationManager.warning("로그인 후 작성가능합니다.", "Error");
          } else {
            NotificationManager.error("오류가 발생하였습니다.", "Error");
          }
          setIsSaving(false);
        }
      }
    },
    [login, isSaving]
  );

  const modifyReply = useCallback(
    async (replyIdx: number, content: string) => {
      if (!isSaving) {
        setIsSaving(true);
        try {
          await replyModify(replyIdx, content).then((res: ResponseType) => {
            if (res.status === 200) {
              NotificationManager.success("댓글을 수정하였습니다.", "Success");
            }
          });
          await getCommentsCallback(Number(idx));
          setIsSaving(false);
        } catch (err) {
          if (err.message === "Error: Request failed with status code 403") {
            NotificationManager.warning("권한이 없습니다.", "Error");
          } else if (err.message === "Error: Request failed with status code 401") {
            cookies.remove("access_token", { path: "/" });
            handleLoginChange(false);
            NotificationManager.warning("로그인 후 작성가능합니다.", "Error");
          } else {
            NotificationManager.error("오류가 발생하였습니다.", "Error");
          }
          setIsSaving(false);
        }
      }
    },
    [login]
  );

  const deleteReplyCallback = useCallback(
    async (replyIdx: number) => {
      try {
        await replyDelete(replyIdx).then((res: ResponseType) => {
          if (res.status === 200) {
            NotificationManager.success("댓글을 삭제하였습니다.", "Success");
          }
        });
        await getCommentsCallback(Number(idx));
        await getPostCommentCount(Number(idx)).then((res: GetPostCommentCountResponse) => {
          setCommentCount(res.data.total_count);
        });
      } catch (err) {
        if (err.message === "Error: Request failed with status code 403") {
          NotificationManager.warning("권한이 없습니다.", "Error");
        } else if (err.message === "Error: Request failed with status code 410") {
          cookies.remove("access_token", { path: "/" });
          NotificationManager.warning("로그인 시간이 만료되었습니다.", "Error");
        } else {
          NotificationManager.error("오류가 발생하였습니다.", "Error");
        }
      }
    },
    [login]
  );

  const deleteReply = useCallback(
    (reply_idx: number) => {
      confirmAlert({
        title: "Warning",
        message: "정말로 삭제하시겠습니까?",
        buttons: [
          {
            label: "Yes",
            onClick: () => deleteReplyCallback(reply_idx)
          },
          {
            label: "No",
            onClick: () => {
              return;
            }
          }
        ]
      });
    },
    [deleteReplyCallback]
  );

  const editPost = useCallback(() => {
    router.push(`/handle/${Number(idx)}`);
  }, [idx]);

  const deletePostCallback = useCallback(() => {
    deletePost(Number(idx))
      .then(() => {
        router.push("/");
        NotificationManager.success("게시글을 삭제하였습니다.", "Success");
      })
      .catch(() => {
        NotificationManager.error("오류가 발생하였습니다.", "Error");
      });
  }, [idx]);

  const deletePostAlert = useCallback(() => {
    confirmAlert({
      title: "Warning",
      message: "정말로 삭제하시겠습니까?",
      buttons: [
        {
          label: "Yes",
          onClick: () => deletePostCallback()
        },
        {
          label: "No",
          onClick: () => setHandler(false)
        }
      ]
    });
  }, [idx]);

  useEffect(() => {
    handlePrevUrl();
    return () => handlePrevUrl();
  }, [prevUrl]);

  useEffect(() => {
    getAllContent();
  }, [getAllContent]);

  useEffect(() => {
    initComments();
    getCommentsCallback(Number(idx));
  }, [getCommentsCallback]);

  return (
    <>
      {(postInfo.idx || (post && post.idx)) && (!postInfo.is_temp || !post.is_temp) && (
        <Head>
          <title>{post.title || postInfo.title}</title>
          <meta
            name="description"
            content={
              post.description
                ? post.description.replace(/ +/g, " ").replace(/#+ |-+ |!+\[+.*\]+\(+.*\)|\`|\>+ |\[!+\[+.*\]+\(+.*\)|\<br+.*\>|\[.*\]\(.*\)/g, "")
                : postInfo.description
                    .replace(/ +/g, " ")
                    .replace(/#+ |-+ |!+\[+.*\]+\(+.*\)|\`|\>+ |\[!+\[+.*\]+\(+.*\)|\<br+.*\>|\[.*\]\(.*\)/g, "") || ""
            }
          />
          <meta property="og:url" content={`https://slog.website/post/${post.idx || postInfo.idx}`} />
          <meta property="og:title" content={post.title || postInfo.title} />
          <meta
            property="og:description"
            content={
              post.description
                ? post.description.replace(/ +/g, " ").replace(/#+ |-+ |!+\[+.*\]+\(+.*\)|\`|\>+ |\[!+\[+.*\]+\(+.*\)|\<br+.*\>|\[.*\]\(.*\)/g, "")
                : postInfo.description
                    .replace(/ +/g, " ")
                    .replace(/#+ |-+ |!+\[+.*\]+\(+.*\)|\`|\>+ |\[!+\[+.*\]+\(+.*\)|\<br+.*\>|\[.*\]\(.*\)/g, "") || ""
            }
          />
          <meta property="twitter:title" content={post.title || postInfo.title} />
          <meta
            property="twitter:description"
            content={
              post.description
                ? post.description.replace(/ +/g, " ").replace(/#+ |-+ |!+\[+.*\]+\(+.*\)|\`|\>+ |\[!+\[+.*\]+\(+.*\)|\<br+.*\>|\[.*\]\(.*\)/g, "")
                : postInfo.description || ""
            }
          />
          {post.title || postInfo.thumbnail ? (
            <>
              <meta property="og:image" content={post.title || postInfo.thumbnail} />
              <meta property="twitter:image" content={post.thumbnail || postInfo.thumbnail} />
            </>
          ) : (
            <>
              <meta property="og:image" content={"https://data.slog.website/public/op_logo.png"} />
              <meta property="twitter:image" content={"https://data.slog.website/public/op_logo.png"} />
            </>
          )}
        </Head>
      )}
      <Post
        commentCount={commentCount}
        deletePostAlert={deletePostAlert}
        handler={handler}
        setHandler={setHandler}
        createComment={createComment}
        modifyComment={modifyComment}
        deleteComment={deleteComment}
        createReply={createReply}
        modifyReply={modifyReply}
        deleteReply={deleteReply}
        userId={userId}
        getReplies={getReplies}
        admin={admin}
        login={login}
        loading={loading}
        comments={comments}
        postInfo={postInfo}
        post={post || null}
        hitPosts={hitPosts}
        editPost={editPost}
      />
    </>
  );
};

export default observer(PostContainer);
