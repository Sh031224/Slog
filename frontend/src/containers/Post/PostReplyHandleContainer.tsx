import PostReplyItem from "components/Post/PostReply/PostReplyItem";
import React, { useCallback, useState } from "react";
import ReplyType from "types/ReplyType";

interface PostReplyHandleContainerProps {
  userId: number;
  admin: boolean;
  reply: ReplyType;
  login: boolean;
  modifyReply: (replyIdx: number, content: string) => Promise<void>;
  deleteReply: (replyIdx: number) => void;
}

const PostReplyHandleContainer = ({
  userId,
  admin,
  reply,
  login,
  modifyReply,
  deleteReply
}: PostReplyHandleContainerProps) => {
  const [modify, setModify] = useState<boolean>(false);
  const [modifyInput, setModifyInput] = useState<string>(reply.content);

  const cancelModify = useCallback(() => {
    setModify(false);
    setModifyInput(reply.content);
  }, [reply]);

  return (
    <>
      <PostReplyItem
        setModifyInput={setModifyInput}
        modifyInput={modifyInput}
        modifyReply={modifyReply}
        deleteReply={deleteReply}
        modify={modify}
        setModify={setModify}
        cancelModify={cancelModify}
        reply={reply}
        admin={admin}
        userId={userId}
        login={login}
      />
    </>
  );
};

export default React.memo(PostReplyHandleContainer);
