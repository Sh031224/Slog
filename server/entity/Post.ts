import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  JoinColumn
} from "typeorm";
import Category from "./Category";

@Entity("post")
export default class Post extends BaseEntity {
  @PrimaryGeneratedColumn()
  idx: number;

  @Column({
    length: 255,
    nullable: false
  })
  title: string;

  @Column({
    length: 255,
    nullable: true
  })
  description: string;

  @Column("text", {
    nullable: false
  })
  content: string;

  @Column({
    nullable: false,
    default: 0
  })
  view: number;

  @Column({
    nullable: false,
    default: false
  })
  is_temp: boolean;

  @ManyToOne((type) => Category, { onDelete: "SET NULL" })
  @JoinColumn({ name: "fk_category_idx" })
  category: Category;

  @Column({
    nullable: true
  })
  fk_category_idx: number;

  @Column({
    length: 1000,
    nullable: true
  })
  thumbnail: string;

  @Column("timestampz")
  @CreateDateColumn()
  created_at: Date;

  @Column("timestampz")
  @CreateDateColumn()
  updated_at: Date;
}
