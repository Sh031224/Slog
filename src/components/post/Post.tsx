import usePostInfo from "hooks/post/usePostInfo";
import React from "react";
import styled from "styled-components";
import Markdown from "../common/markdown/Markdown";
import PostComment from "./comment/PostComment";
import PostHeader from "./content/PostHeader";
import PostLoading from "./content/PostLoading";
import PostHit from "./hit/PostHit";

const Post: React.FC = () => {
  const { data, loading } = usePostInfo();
  const { post } = data;

  return (
    <PostWrapper>
      <PostContainer>
        {loading ? (
          <PostLoading />
        ) : (
          <>
            <PostHeader />
            <PostContentContainer>
              <Markdown content={post.content} />
            </PostContentContainer>
            <PostHit />
            <PostComment />
          </>
        )}
      </PostContainer>
    </PostWrapper>
  );
};

const PostWrapper = styled.main`
  width: 100%;
  display: flex;
  justify-content: center;
`;

const PostContainer = styled.div`
  max-width: 900px;
  position: relative;
  width: 100%;
  padding: 0 1rem;
`;

const PostContentContainer = styled.section`
  margin: 3rem 0;
`;

export default Post;
