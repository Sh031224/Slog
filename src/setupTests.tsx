import { ThemeProvider } from "styled-components";
import { render } from "@testing-library/react";
import { lightTheme } from "style/theme";
import { store } from "stores";
import { Provider } from "react-redux";
import { NotificationContainer } from "react-notifications";

const Providers = ({ children }) => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={lightTheme}>
        <NotificationContainer />
        {children}
      </ThemeProvider>
    </Provider>
  );
};

const customRender = (ui, options = {}) => render(ui, { wrapper: Providers, ...options });

// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom/extend-expect";
import "jest-styled-components";

(window as any).scrollTo = () => {};

export { customRender as render };
