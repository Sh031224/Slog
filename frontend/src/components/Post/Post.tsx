import React from "react";
import CommentType from "types/CommentType";
import { PostInfoType, PostType } from "types/PostType";
import { GetRepliesResponse } from "types/Response";
import MarkdownContainer from "../../containers/Markdown/MarkdownContainer";
import "./Post.scss";
import PostComment from "./PostComment";
import PostHeader from "./PostHeader";
import PostHit from "./PostHit";
import PostLoading from "./PostLoading";

interface PostProps {
  loading: boolean;
  comments: CommentType[];
  post: PostInfoType;
  postInfo: PostInfoType | null;
  hitPosts: PostType[];
  login: boolean;
  admin: boolean;
  createComment: (
    postIdx: number,
    content: string,
    isPrivate?: boolean | undefined
  ) => Promise<void>;
  getReplies: (commentIdx: number) => Promise<GetRepliesResponse>;
  userId: number;
  modifyComment: (commentIdx: number, content: string) => Promise<void>;
  deleteComment: (commentIdx: number) => void;
  createReply: (
    commentIdx: number,
    content: string,
    isPrivate?: boolean | undefined
  ) => Promise<void>;
  modifyReply: (reply_idx: number, content: string) => Promise<void>;
  deleteReply: (reply_idx: number) => void;
  handler: boolean;
  setHandler: React.Dispatch<React.SetStateAction<boolean>>;
  deletePostAlert: () => void;
  editPost: () => void;
  commentCount: number;
}

const Post = ({
  loading,
  comments,
  post,
  postInfo,
  hitPosts,
  login,
  admin,
  createComment,
  getReplies,
  userId,
  modifyComment,
  deleteComment,
  createReply,
  modifyReply,
  deleteReply,
  handler,
  setHandler,
  deletePostAlert,
  editPost,
  commentCount
}: PostProps) => {
  return (
    <div className="post">
      <div className="post-box">
        {!post.idx && loading ? (
          <PostLoading />
        ) : (
          <>
            <PostHeader
              editPost={editPost}
              deletePostAlert={deletePostAlert}
              handler={handler}
              setHandler={setHandler}
              admin={admin}
              thumbnail={post.thumbnail || postInfo.thumbnail}
              title={post.title || postInfo.title}
              createdAt={post.created_at || postInfo.created_at}
              updatedAt={post.updated_at || postInfo.updated_at}
            />

            <MarkdownContainer className="post-content">
              {post.content || postInfo.content}
            </MarkdownContainer>
            {!loading && (
              <>
                <PostHit hitPosts={hitPosts} />
                <PostComment
                  commentCount={commentCount}
                  createComment={createComment}
                  deleteComment={deleteComment}
                  modifyComment={modifyComment}
                  createReply={createReply}
                  modifyReply={modifyReply}
                  deleteReply={deleteReply}
                  userId={userId}
                  getReplies={getReplies}
                  postIdx={post.idx || postInfo.idx}
                  admin={admin}
                  login={login}
                  comments={comments}
                />
              </>
            )}
          </>
        )}
      </div>
    </div>
  );
};

export default Post;
