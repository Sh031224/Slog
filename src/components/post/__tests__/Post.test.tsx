import { render } from "setupTests";
import Post from "../Post";

jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: "/post/1",
      pathname: "",
      query: { idx: 1 },
      asPath: "",
      push: jest.fn()
    };
  }
}));

describe("Post", () => {
  const setup = (props = {}) => {
    return render(<Post {...props} />);
  };

  it("renders correctly", () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });
});
