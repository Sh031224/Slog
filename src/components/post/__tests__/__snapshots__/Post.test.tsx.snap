// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post renders correctly 1`] = `
.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  margin-top: 5rem;
  -webkit-animation: loading 3s infinite ease-in-out;
  animation: loading 3s infinite ease-in-out;
  width: 100%;
  height: 3.4rem;
  border-radius: 5px;
}

.c4 {
  width: 10rem;
  border-radius: 5px;
  margin-top: 1rem;
  -webkit-animation: loading 3s infinite ease-in-out;
  animation: loading 3s infinite ease-in-out;
  height: 2rem;
}

.c5 {
  width: 100%;
  height: 25rem;
  -webkit-animation: loading 3s infinite ease-in-out;
  animation: loading 3s infinite ease-in-out;
  border-radius: 5px;
  margin-top: 5rem;
}

.c6 {
  margin-top: 1rem;
  height: 3.4rem;
  border-radius: 5px;
  width: 100%;
  -webkit-animation: loading 3s infinite ease-in-out;
  animation: loading 3s infinite ease-in-out;
}

.c7 {
  margin-top: 1rem;
  height: 3.4rem;
  border-radius: 5px;
  width: 60%;
  -webkit-animation: loading 3s infinite ease-in-out;
  animation: loading 3s infinite ease-in-out;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  max-width: 900px;
  position: relative;
  width: 100%;
  padding: 0 1rem;
}

@media only screen and (max-width:767px) {

}

@media only screen and (max-width:767px) {

}

@media only screen and (max-width:767px) {

}

@media only screen and (max-width:767px) {

}

@media only screen and (max-width:767px) {

}

@media only screen and (max-width:767px) {

}

@media only screen and (max-width:767px) {

}

@media only screen and (max-width:767px) {

}

@media only screen and (max-width:900px) {

}

@media only screen and (max-width:900px) {

}

@media only screen and (max-width:500px) {

}

@media only screen and (max-width:900px) {

}

@media only screen and (max-width:900px) {
  .c3 {
    margin-top: 1rem;
    height: 3rem;
  }
}

@media only screen and (max-width:900px) {
  .c5 {
    height: 20rem;
    margin-top: 3rem;
  }
}

@media only screen and (max-width:500px) {
  .c5 {
    height: 18rem;
  }
}

@media only screen and (max-width:900px) {
  .c6 {
    height: 3rem;
  }
}

@media only screen and (max-width:900px) {
  .c7 {
    height: 3rem;
  }
}

<div>
  <div
    class="notification-container notification-container-empty"
  >
    <div />
  </div>
  <main
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        />
        <div
          class="c4"
        />
        <div
          class="c5"
        />
        <div
          class="c6"
          width="100%"
        />
        <div
          class="c7"
          width="60%"
        />
      </div>
    </div>
  </main>
</div>
`;
